zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+/// already used. If `reject_all_ephemeral` is set, return an error if the address
+/// is ephemeral at all, regardless of reuse.
+fn ephemeral_address_check_internal<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    address_str: &str,
+    reject_all_ephemeral: bool,
+) -> Result<(), SqliteClientError> {
+    // It is intentional that we don't require `t.mined_height` to be non-null.
+    // That is, we conservatively treat an ephemeral address as potentially
+    // reused even if we think that the transaction where we had evidence of
+    // its use is at present unmined. This should never occur in supported
+    // situations where only a single correctly operating wallet instance is
+    // using a given seed, because such a wallet will not reuse an address that
+    // it ever reserved.
+    //
+    // `COALESCE(used_in_tx, mined_in_tx)` can only differ from `used_in_tx`
+    // if the address was reserved, an error occurred in transaction creation
+    // before calling `mark_ephemeral_address_as_used`, and then we observed
+    // the address to have been used in a mined transaction (presumably by
+    // another wallet instance, or due to a bug) anyway.
+    let res = wdb
+        .conn
+        .0
+        .query_row(
+            "SELECT t.txid FROM ephemeral_addresses
+             LEFT OUTER JOIN transactions t
+             ON t.id_tx = COALESCE(used_in_tx, mined_in_tx)
+             WHERE address = :address",
+            named_params![":address": address_str],
+            |row| row.get::<_, Option<Vec<u8>>>(0),
+        )
+        .optional()?;
+
+    match res {
+        Some(Some(txid_bytes)) => {
+            let txid = TxId::from_bytes(
+                txid_bytes
+                    .try_into()
+                    .map_err(|_| SqliteClientError::CorruptedData("invalid txid".to_owned()))?,
+            );
+            Err(SqliteClientError::EphemeralAddressReuse(
+                address_str.to_owned(),
+                Some(txid),
+            ))
+        }
+        Some(None) if reject_all_ephemeral => Err(SqliteClientError::EphemeralAddressReuse(
+            address_str.to_owned(),
+            None,
+        )),
+        _ => Ok(()),
+    }
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses.
+///
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used.
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, ephemeral_address);
+    ephemeral_address_check_internal(wdb, &address_str, false)?;
```

Comment:
blocking: This is a deeply weird place to be making this check, and the inability to mark an address as used could potentially brick the user's wallet. I'm not sure what the right thing to do is in this circumstance, but we can't depend upon this error being raised at a point that doesn't leave the user's wallet in an inconsistent state.

Reply:
If `mark_ephemeral_address_as_used` is called more than once for each address, it indicates a programming error specifically in `zcash_client_backend` or `zcash_client_sqlite`. We allocate an address by the call to `reserve_next_n_ephemeral_addresses` in `create_proposed_transaction`. At that point it has been permanently placed into the `ephemeral_addresses` table, and should never be returned by any subsequent call to `reserve_next_n_ephemeral_addresses`. We then make sure that the address is not leaked other than to use it in the transaction being constructed, until we call `store_sent_tx` in the same sequential flow of execution. Note that this is not dependent on any heuristics about gap limits or the behaviour of other wallet instances (including ones that share the seed).

An error here doesn't block marking an address as used in the database, because it was already marked as used in order for the error to occur. All it does is make sure than an unsafe situation due to a programming error is reported. It couldn't brick the user's wallet, because `mark_ephemeral_address_as_used` is only called in the `Recipient::EphemeralTransparent` case for the recipient of an output in `sent_tx.outputs()`. So it can only occur when calling `store_sent_tx` after creating a transaction that makes use of ephemeral transparent addresses (i.e. currently, a ZIP 320 transaction).

If you create `2^{31}` ZIP 320 transactions, then that will brick the ability to create more ZIP 320 transactions in that account, but that's designed behaviour, and should not happen within a reasonable wallet lifetime.